src/net/ossl.cc

```diff
@@ -146,6 +146,10 @@ std::system_error make_ossl_error(const std::string & msg) {
     }
 }
 
+inline void drain_openssl_error_stack() {
+    while (ERR_get_error() != 0) {}
+}
```

Comment:
q: is there any observable effect when we _don't_ do this? i assume it wouldn't rise to the level of a memory leak, but I'm curious.

Reply:
The only issue may be possible user confusion when we look at the errors coming out of OpenSSL (if we have to look at them at all).
Also turns out [`ERR_clear_error`](https://www.openssl.org/docs/man3.0/man3/ERR_clear_error.html)... 💁

Reply:
`make_ossl_error` returns `front` into the `system_error`, which would be the earliest, so not clearing it could cause incorrect handling? I'm almost tempted to suggest calling it in a `ss::defer` in all the error handling blocks.

Reply:
> I'm almost tempted to suggest calling it in a ss::defer in all the error handling blocks.

Sorry wdym?

Reply:
> > I'm almost tempted to suggest calling it in a ss::defer in all the error handling blocks.
> 
> Sorry wdym?

When a function returns an error_code, and it's handled, always clear the ssl errors.

I think `ss::defer` should work since when the error is inspected with `ERR_get_error` they all do so synchronously, and then all exits from the scope would then clear the errors, leaving a clean slate for the next time an error occurred.

Reply:
Examining the error is more of the exception to the rule.  Most of the time we just see if the function errored, not why.  Getting the answer to "why" requires the ERR_peak call as nearly all functions in OpenSSL just return a 1 for success or 0 for failure.

Reply:
> Examining the error is more of the exception to the rule. Most of the time we just see if the function errored, not why. Getting the answer to "why" requires the ERR_peak call as nearly all functions in OpenSSL just return a 1 for success or 0 for failure.

OK, so imagine this:
1. A call fails
2. We don't clear out the error
3. A call fails in an unrelated context
4. We inspect the error with `make_ossl_error`

The `system_error` now contains the unrelated error code from the failure in 1.

Or am I misunderstanding?

Reply:
> Or am I misunderstanding?

No you are, but I think we're talking past each other :)

The vast majority of failures results in a call to `make_ossl_error`.  I have gone through and checked that any failure that doesn't and instead performs an error code stack inspection for 'special handling' calls `ERR_clear_error` so the above scenario _doesn't_ happen.